### PR TITLE
[records] Allow "()" as an empty record expression.

### PR DESCRIPTION
**Pull request by munificent**
I believe we decided to do this in a previous language meeting, just catching the proposal up to that.

### CODE REVIEW

**Review by leafpetersen — APPROVED** [review]

**Comment by leafpetersen** [comment]
@chloestefantsova @johnniwinther @scheglov FYI, `Record.empty` goes away, replaced with `()`.

**Comment by munificent** [comment]
> FYI, `Record.empty` goes away, replaced with `()`.

That reminds me, I should update the CHANGELOG.

**Comment by johnniwinther** [comment]
cc @jensjoha

File: accepted/future-releases/records/records-feature-specification.md [hunk 1]
```diff
@@ -122,6 +122,8 @@ not captured by the grammar. It is a compile-time error if a record has any of:
 
 *   Only one positional field and no trailing comma.
 
+*   No fields and a trailing comma. *The expression `(,)` isn't allowed.*
```
Comment (lrhn) [inline accepted/future-releases/records/records-feature-specification.md:125]: My guess is that this is (or should be) captured by the grammar. 
In the rest of our grammar, an optional trailing colon is usually handled as:
```dart
collection ::= startToken elements? endToken
elements ::= element (',' element)* ','?
```

The grammar above doesn't actually allow `()` as a record. And it allows `(e)`, which is ambiguous.
Having a prose section saying that it isn't allowed is technically enough, but not how we usually do it.

How about:
```abnf
record ::= 'const'? '(' recordFields? ')'

recordFields ::= expression (',' | (',' recordField)+ ','?)
               | namedRecordField ( ',' recordField )* ','?

recordField ::= expression | namedRecordField
namedRecordField ::= identifier ':' expression
```

An alternative phrasing would be:
```dart
recordFields ::= expression ','
               | namedRecordField ','?,
               | recordField (',' recordField)+ ','?
```

Comment (munificent) [reply accepted/future-releases/records/records-feature-specification.md:125]: > My guess is that this is (or should be) captured by the grammar.

I thought about that.

> Having a prose section saying that it isn't allowed is technically enough, but not how we usually do it.

Yes, and the implementers tell me often that they don't like the way we currently do it. :)

Every time I try to sort out `type`, `typeNotVoid`, `typeNotFunction`, and `typeNotVoidNotFunction`, I get a little sadder.

**Pull request merged**